README.md
# holdem_sim

Get probablilites, odds, and win expectations for holdem hands

## Installation

```bash
$ pip install holdem_sim
```

## Usage

Read the docs at [readthedocs.org/holdem-sim](https://readthedocs.org/projects/holdem-sim/)

## Credits

`holdem_sim` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.1.0 (22/11/2023)

- Deprecate the `main` module and rename it `holdem`.  Move Pretty table functions to new `sim_results` module.


## v0.1.0 (17/05/2023)

- First release of `holdem_sim`!

tests/test_holdem_sim.py
from holdem_sim import holdem_sim


tests/test_simulation.py
import pytest
import holdem_sim.poker_functions
import holdem_sim.simulation as s

@pytest.fixture
def six_card_straight_board():
    # Straight, 6 card
    flop9 = ['Jh', '3c', '6d']
    turn9 = ['4h']
    river9 = ['5s']
    board = flop9 + turn9 + river9
    return board


@pytest.fixture
def six_card_straight_hand():
    hand9 = ['2c', '7s']
    return hand9

#  Duplicated Card
@pytest.fixture
def duplicate_hand():
    hand10 = ['3c', 'As']
    flop10 = ['3c', 'Jd', '9h']
    turn10 = ['4s']
    river10 = ['8s']
    check = hand10 + flop10 + turn10 + river10
    return check


@pytest.fixture
def impossible_straight():
    hand = ['As', 'Kd']
    flop = ['Kh', '6c', '4s']
    return hand, flop


#  Invalid Card
@pytest.fixture
def invalid_card():
    hand11 = ['3c', 'As']
    flop11 = ['9c', 'Jd', '9h']
    turn11 = ['4s']
    river11 = ['Ss']
    check = hand11 + flop11 + turn11 + river11
    return check

def test_simulation_incomplete_board():
    """Will the sim run without errors with an incomplete board? Minimum result is 1 pair"""
    hand = ['Ac', '3d']
    flop = ['As', '5c', '4d']
    sims = 5
    sim = s.simulation_one_player(hand, flop, sims=sims)
    assert sim[0] == 5


def test_no_impossible_straight(impossible_straight):
    """Prior to refactor, this hand and others similar would yield a small % of straights"""
    hand = impossible_straight[0]
    flop = impossible_straight[1]
    sim = s.simulation_one_player(hand, flop)
    assert sim[5] == 0


def test_duplicate_card(duplicate_hand):
    check = duplicate_hand
    duplicate = s.dedupe(check)
    assert duplicate


def test_not_duplicate(six_card_straight_hand, six_card_straight_board):
    check = six_card_straight_hand + six_card_straight_board
    duplicate = s.dedupe(check)
    assert not duplicate


def test_valid_card(six_card_straight_hand, six_card_straight_board):
    check = six_card_straight_hand + six_card_straight_board
    valid = s.validate_card(check)
    assert valid


def test_invalid_card(invalid_card):
    check = invalid_card
    valid = s.validate_card(check)
    assert not valid


def test_player_create():
    player = s.Player(1, ['Ac', 'Ad'])
    assert type(player) == s.Player


def test_player_cards():
    player = s.Player(1, ['Ac', 'Ad'])
    assert type(player.cards[0]) == holdem_sim.poker_functions.Card


def test_player_no_cards():
    player = s.Player(2)
    assert player

def test_multiplayer_create_players_no_hole_cards():
    foo = s.simulation_multiplayer(['As', '9d'], opponents=3)
    assert len(foo) == 3


def test_multiplayer_create_players_with_hole_cards():
    foo = s.simulation_multiplayer(['As', '9d'], ['Jd', '8c'], ['6h', 'Kh'], opponents=3)
    assert len(foo) == 3


def test_multiplayer_with_hole_cards():
    foo = s.simulation_multiplayer(['As', '9d'], ['Kd', 'Th'], opponents=2)
    assert len(foo[1].cards) == 2


def test_score_game_single_winner():
    player0 = s.Player(0)
    player1 = s.Player(1)
    player2 = s.Player(2)

    player0.hand = holdem_sim.poker_functions.Hand('3ok', 8, kicker=10)
    player1.hand = holdem_sim.poker_functions.Hand('hc', 13, 10, 9)
    player2.hand = holdem_sim.poker_functions.Hand('pair', 8, kicker=12)

    contestants = [player0, player1, player2]

    foo = s.score_game(contestants)

    assert foo[0].wins == 1


def test_score_game_high_winner():
    """More than 1 hand has a straight, flush or straight flush.  High card wins"""
    player0 = s.Player(0)
    player1 = s.Player(1)
    player2 = s.Player(2)

    player0.hand = holdem_sim.poker_functions.Hand('flush', 13)
    player1.hand = holdem_sim.poker_functions.Hand('flush', 9)
    player2.hand = holdem_sim.poker_functions.Hand('pair', 8, kicker=12)

    contestants = [player0, player1, player2]

    foo = s.score_game(contestants)

    assert foo[0].wins == 1


def test_score_game_high_chop():
    """More than 1 hand has a straight, flush or straight flush.  Draw."""
    player0 = s.Player(0)
    player1 = s.Player(1)
    player2 = s.Player(2)

    player0.hand = holdem_sim.poker_functions.Hand('flush', 13)
    player1.hand = holdem_sim.poker_functions.Hand('flush', 13)
    player2.hand = holdem_sim.poker_functions.Hand('pair', 8, kicker=12)

    contestants = [player0, player1, player2]

    foo = s.score_game(contestants)

    assert foo[0].wins == 0 and foo[1].wins == 0


def test_score_game_boat_high_plays():
    """More than 1 hand has a straight, flush or straight flush.  High card wins"""
    player0 = s.Player(0)
    player1 = s.Player(1)
    player2 = s.Player(2)

    player0.hand = holdem_sim.poker_functions.Hand('boat', 13, 5)
    player1.hand = holdem_sim.poker_functions.Hand('boat', 9, 6)
    player2.hand = holdem_sim.poker_functions.Hand('pair', 8, kicker=12)

    contestants = [player0, player1, player2]

    foo = s.score_game(contestants)

    assert foo[0].wins == 1


def test_score_game_boat_low_plays():
    """More than 1 hand has a straight, flush or straight flush.  High card wins"""
    player0 = s.Player(0)
    player1 = s.Player(1)
    player2 = s.Player(2)

    player0.hand = holdem_sim.poker_functions.Hand('boat', 13, 5)
    player1.hand = holdem_sim.poker_functions.Hand('boat', 13, 6)
    player2.hand = holdem_sim.poker_functions.Hand('pair', 8, kicker=12)

    contestants = [player0, player1, player2]

    foo = s.score_game(contestants)

    assert foo[1].wins == 1



def test_score_game_hi_lo_kick_clear_hi_winner():
    """More than 1 hand has a straight, flush or straight flush.  High card wins"""
    player0 = s.Player(0)
    player1 = s.Player(1)
    player2 = s.Player(2)

    player0.hand = holdem_sim.poker_functions.Hand('boat', 13, 5)
    player1.hand = holdem_sim.poker_functions.Hand('boat', 10, 6)
    player2.hand = holdem_sim.poker_functions.Hand('pair', 8, kicker=12)

    contestants = [player0, player1, player2]

    foo = s.score_game(contestants)

    assert foo[0].wins == 1


def test_score_game_hi_lo_kick_lo_winner():
    """More than 1 hand has trips.  Fourth card plays"""
    player0 = s.Player(0)
    player1 = s.Player(1)
    player2 = s.Player(2)

    player0.hand = holdem_sim.poker_functions.Hand('3ok', 13, 5, 3)
    player1.hand = holdem_sim.poker_functions.Hand('3ok', 10, 6, 5)
    player2.hand = holdem_sim.poker_functions.Hand('pair', 8, kicker=12)

    contestants = [player0, player1, player2]

    foo = s.score_game(contestants)

    assert foo[0].wins == 1


def test_score_game_hi_lo_kick_kicker_winner():
    """More than 1 hand has trips.  Fifth card plays """
    player0 = s.Player(0)
    player1 = s.Player(1)
    player2 = s.Player(2)

    player0.hand = holdem_sim.poker_functions.Hand('3ok', 13, 5, 3)
    player1.hand = holdem_sim.poker_functions.Hand('3ok', 13, 5, 4)
    player2.hand = holdem_sim.poker_functions.Hand('pair', 8, kicker=12)

    contestants = [player0, player1, player2]

    foo = s.score_game(contestants)

    assert foo[1].wins == 1


@pytest.fixture
def quad_showdown():
    """More than 1 hand has quads.  Fifth card plays """
    player0 = s.Player(0)
    player1 = s.Player(1)
    player2 = s.Player(2)

    player0.hand = holdem_sim.poker_functions.Hand('4ok', 5, low_value=3)
    player1.hand = holdem_sim.poker_functions.Hand('4ok', 5, low_value=4)
    player2.hand = holdem_sim.poker_functions.Hand('pair', 8, low_value=12, kicker=9)

    contestants = [player0, player1, player2]
    return contestants


def test_score_game_kicker_winner(quad_showdown):

    contestants = quad_showdown

    foo = s.score_game(contestants)

    assert foo[1].wins == 1

tests/test_poker_functions.py
import pytest
import src.holdem_sim.poker_functions as p

# valid card string
card_str1 = 'As'

# invalid card string
card_string2 = 'Of'

#  hand with kicker
new_hand = p.Hand('pair', 12, 11)

@pytest.fixture
def two_pair():
    # 2 Pair
    hand1 = ['As', '4c']
    flop1 = ['4d', 'Qh', 'Ts']
    turn1 = ['7h']
    river1 = ['Ad']
    board = flop1 + turn1 + river1
    return hand1, board


@pytest.fixture
def flush_one_pair():
    # Flush, 1 Pair
    hand2 = ['As', 'Ks']
    flop2 = ['4d', 'Kh', 'Ts']
    turn2 = ['7s']
    river2 = ['3s']
    board = flop2 + turn2 + river2
    return hand2, board


@pytest.fixture
def straight_no_pair_no_flush():
    # Straight, no pair, no flush
    hand3 = ['4s', '5d']
    flop3 = ['6c', '7s', '8h']
    turn3 = ['Jc']
    river3 = ['As']
    board = flop3 + turn3 + river3
    return hand3, board


@pytest.fixture
def three_of_a_kind():
    # 3 of a kind
    hand4 = ['As', 'Ac']
    flop4 = ['4d', 'Qh', 'Ts']
    turn4 = ['7h']
    river4 = ['Ad']
    board = flop4  + turn4 + river4
    return hand4, board


@pytest.fixture
def boat():
    # Full House
    hand5 = ['As', 'Ac']
    flop5 = ['4d', 'Qh', 'Ts']
    turn5 = ['4c']
    river5 = ['Ad']
    board = flop5 + turn5 + river5
    return hand5, board



@pytest.fixture
def straight_flush():
    # Straight flush
    hand6 = ['4s', '5s']
    flop6 = ['6s', '7s', '8s']
    turn6 = ['Jc']
    river6 = ['Ac']
    board = flop6 + turn6 + river6
    return hand6, board

@pytest.fixture
def non_seq_straight():
    # Straight, non-sequential
    hand7 = ['2c', '9s']
    flop7 = ['Jh', '6c', '3d']
    turn7 = ['4h']
    river7 = ['5s']
    board = flop7 + turn7 + river7
    return hand7, board


@pytest.fixture
def straight_five_card():
    # Straight, 5 card
    hand8 = ['2c', '2s']
    flop8 = ['6h', '3c', '3d']
    turn8 = ['4h']
    river8 = ['5s']
    board = flop8 + turn8 + river8
    return hand8, board

@pytest.fixture
def straight_six_card():
# Straight, 6 card
    hand = ['2c', '7s']
    flop9 = ['Jh', '3c', '6d']
    turn9 = ['4h']
    river9 = ['5s']
    board = flop9 + turn9 + river9
    return hand, board

#  Duplicated Card
hand10 = ['3c', 'As']
flop10 = ['3c', 'Jd', '9h']
turn10 = ['4s']
river10 = ['8s']


#  Invalid Card
hand11 = ['3c', 'As']
flop11 = ['9c', 'Jd', '9h']
turn11 = ['4s']
river11 = ['Ss']


@pytest.fixture
def wheel():
    #  Wheel
    hand12 = ['Ac', 'Td']
    flop12 = ['2c', '3h', 'Qs']
    turn12 = ['4s']
    river12 = ['5d']
    board = flop12 + turn12 + river12
    return  hand12, board


@pytest.fixture
def high_card():
    #  High Card
    hand13 = ['Td', '4c']
    flop13 = ['2c', '3h', 'Qs']
    turn13 = ['5d']
    river13 = ['9c']
    board = flop13 + turn13 + river13
    return hand13, board


@pytest.fixture
def quads():
    hand14 = ['4c', '4h']
    flop14 = ['4s', 'Ac', '9h']
    turn14 = ['8d']
    river14 = ['4d']
    board = flop14 + turn14 + river14
    return hand14, board

def test_card_init():
    card = p.Card(card_str1)
    assert type(card) == p.Card


def test_card_rank():
    card = p.Card(card_str1)
    assert card.rank == 'A'


def test_card_suit():
    card = p.Card(card_str1)
    assert card.suit == 's'


def test_card_value():
    card = p.Card(card_str1)
    assert card.value == 14


def test_card_name():
    card = p.Card(card_str1)
    assert card.name == 'As'


def test_deck_init():
    deck = p.generate_deck()
    assert type(deck) == p.Deck


def test_deck_contents():
    deck = p.generate_deck()
    assert type(deck[5]) == p.Card


def test_deck_size():
    deck = p.generate_deck()
    assert len(deck) == 52


def test_deal_card_1():
    """Ensures that the deal_card generates a Card object"""
    deck = p.generate_deck()
    card, deck = deck.deal_card()
    assert type(card) == p.Card


def test_deal_card_2():
    """Ensures that deal_card removes Card from deck"""
    deck = p.generate_deck()
    card, deck = deck.deal_card()
    assert len(deck) == 51


def test_update_deck_1():
    """Remove a passed card from Deck passed as string"""
    deck = p.generate_deck()
    deck.update_deck('Ks')
    cards = [card.name for card in deck]
    assert len(deck) == 51


def test_update_deck_2():
    """Remove a passed card from Deck passed as Card"""
    card = p.Card('Ks')
    deck = p.generate_deck()
    deck.update_deck(card)
    cards = [card.name for card in deck]
    assert len(deck) == 51


def test_update_deck_3():
    """Ensure specific passed card is removed from deck"""
    passed_card = p.Card('2c')
    deck = p.generate_deck()
    deck.update_deck(passed_card)
    cards = [card.name for card in deck]
    assert passed_card.name not in cards


def test_update_deck_4():
    """Ensure specific passed card string is removed from deck"""
    passed_card = '2c'
    deck = p.generate_deck()
    deck.update_deck(passed_card)
    cards = [card.name for card in deck]
    assert passed_card not in cards


def test_no_flush(two_pair):
    board = two_pair[1]
    hand = two_pair[0]
    flush = p.find_flush(hand, board)
    assert not flush


def test_flush_hand_value(flush_one_pair):
    board2 = flush_one_pair[1]
    hand = flush_one_pair[0]
    flush_hand = p.find_flush(hand, board2)
    assert flush_hand.high_value == 14


def test_flush_hand_type(flush_one_pair):
    board2 = flush_one_pair[1]
    hand = flush_one_pair[0]
    flush_hand = p.find_flush(hand, board2)
    assert flush_hand.type == 'flush'

def test_one_pair_type(flush_one_pair):
    board2 = flush_one_pair[1]
    hand = flush_one_pair[0]
    multiple_hand = p.find_multiple(hand, board2)
    assert multiple_hand.type == 'pair'


def test_one_pair_high_value(flush_one_pair):
    board2 = flush_one_pair[1]
    hand = flush_one_pair[0]
    multiple_hand = p.find_multiple(hand, board2)
    assert multiple_hand.high_value == 13


def test_one_pair_low_value(flush_one_pair):
    board2 = flush_one_pair[1]
    hand = flush_one_pair[0]
    multiple_hand = p.find_multiple(hand, board2)
    assert multiple_hand.low_value == 14


def test_one_pair_kicker_value(flush_one_pair):
    board2 = flush_one_pair[1]
    hand = flush_one_pair[0]
    multiple_hand = p.find_multiple(hand, board2)
    assert multiple_hand.kicker == 10

def test_not_one_pair(straight_no_pair_no_flush):
    board = straight_no_pair_no_flush[1]
    hand = straight_no_pair_no_flush[0]
    pair = p.find_pair(hand, board)
    assert not pair

def test_not_two_pair(straight_no_pair_no_flush):
    board3 = straight_no_pair_no_flush[1]
    hand3 = straight_no_pair_no_flush[0]
    two_pair = p.find_two_pair(hand3, board3)
    assert not two_pair

def test_two_pair_type(two_pair):
    board1 = two_pair[1]
    hand1 = two_pair[0]
    two_pair_hand = p.find_two_pair(hand1, board1)
    assert two_pair_hand.type == '2pair' and two_pair_hand.kicker_rank == 'Q'



def test_two_pair_high(two_pair):
    board1 = two_pair[1]
    hand1 = two_pair[0]
    two_pair_hand = p.find_two_pair(hand1, board1)
    assert two_pair_hand.high_value == 14


def test_two_pair_low(two_pair):
    board1 = two_pair[1]
    hand1 = two_pair[0]
    two_pair_hand = p.find_two_pair(hand1, board1)
    assert two_pair_hand.low_value == 4

def test_not_3ok(two_pair):
    board1 = two_pair[1]
    hand1 = two_pair[0]
    three_o_kind = p.find_trips(hand1, board1)
    assert not three_o_kind


def test_3ok_type(three_of_a_kind):
    board = three_of_a_kind[1]
    hand = three_of_a_kind[0]
    three_o_kind_hand = p.find_trips(hand, board)
    assert three_o_kind_hand.type == '3ok'


def test_3ok_high_value(three_of_a_kind):
    board = three_of_a_kind[1]
    hand = three_of_a_kind[0]
    three_o_kind_hand = p.find_trips(hand, board)
    assert three_o_kind_hand.high_value == 14


def test_3ok_low_value(three_of_a_kind):
    board = three_of_a_kind[1]
    hand = three_of_a_kind[0]
    three_o_kind_hand = p.find_trips(hand, board)
    assert three_o_kind_hand.low_value == 12


def test_3ok_kicker(three_of_a_kind):
    board = three_of_a_kind[1]
    hand = three_of_a_kind[0]
    three_o_kind_hand = p.find_trips(hand, board)
    assert three_o_kind_hand.kicker == 10


def test_straight_sequential(straight_no_pair_no_flush):
    board3 = straight_no_pair_no_flush[1]
    hand3 = straight_no_pair_no_flush[0]
    straight_hand = p.find_straight(hand3, board3)
    assert straight_hand.type == 'straight'

def test_straight_non_sequential(non_seq_straight):
    board7 = non_seq_straight[1]
    hand7 = non_seq_straight[0]
    straight_hand = p.find_straight(hand7, board7)
    assert straight_hand.high_rank == '6'


def test_straight_5_card_type(straight_five_card):
    board8 = straight_five_card[1]
    hand8 = straight_five_card[0]
    straight_hand = p.find_straight(hand8, board8)
    assert straight_hand.type == 'straight'


def test_straight_5_card_high(straight_five_card):
    board8 = straight_five_card[1]
    hand8 = straight_five_card[0]
    straight_hand = p.find_straight(hand8, board8)
    assert straight_hand.high_value == 6


def test_straight_5_card_low(straight_five_card):
    board8 = straight_five_card[1]
    hand8 = straight_five_card[0]
    straight_hand = p.find_straight(hand8, board8)
    assert straight_hand.low_value == 0


def test_straight_5_card_kicker(straight_five_card):
    board8 = straight_five_card[1]
    hand8 = straight_five_card[0]
    straight_hand = p.find_straight(hand8, board8)
    assert straight_hand.kicker == 0


def test_straight_6_card(straight_six_card):
    board9 = straight_six_card[1]
    hand9 = straight_six_card[0]
    straight_hand = p.find_straight(hand9, board9)
    assert straight_hand.high_value == 7


def test_not_straight(two_pair):
    board4 = two_pair[1]
    hand4 = two_pair[0]
    straight = p.find_straight(hand4, board4)
    assert not straight


def test_straight_wheel(wheel):
    board12 = wheel[1]
    hand12 = wheel[0]
    straight_hand = p.find_straight(hand12, board12)
    assert straight_hand.high_rank == '5'


def test_not_straight_flush(boat):
    board = boat[1]
    hand = boat[0]
    straight_flush = p.find_straight_flush(hand, board)
    assert not straight_flush

def test_not_straight_flush_flush(flush_one_pair):
    board = flush_one_pair[1]
    hand = flush_one_pair[0]
    straight_flush = p.find_straight_flush(hand, board)
    assert not straight_flush


def test_straight_flush(straight_flush):
    board = straight_flush[1]
    hand = straight_flush[0]
    straight_flush_hand = p.find_straight_flush(hand, board)
    assert straight_flush_hand.high_value == 8

def test_hand_init():
    my_hand = new_hand
    assert type(my_hand) == p.Hand

def test_hand_value():
    my_hand = new_hand
    assert new_hand.hand_value == 2


def test_high_card_type(high_card):
    board = high_card[1]
    hand = high_card[0]
    high_card_hand = p.find_high_card(hand, board)
    assert high_card_hand.type == 'hc'


def test_high_card_high(high_card):
    board = high_card[1]
    hand = high_card[0]
    high_card_hand = p.find_high_card(hand, board)
    assert high_card_hand.high_value == 12


def test_high_card_low(high_card):
    board = high_card[1]
    hand = high_card[0]
    high_card_hand = p.find_high_card(hand, board)
    assert high_card_hand.low_value == 10


def test_high_card_kicker(high_card):
    board = high_card[1]
    hand = high_card[0]
    high_card_hand = p.find_high_card(hand, board)
    assert high_card_hand.kicker == 9


def test_quads_type(quads):
    board = quads[1]
    hand = quads[0]
    quads = p.find_quads(hand, board)
    assert quads.type == '4ok'


def test_quads_high(quads):
    board = quads[1]
    hand = quads[0]
    quads = p.find_quads(hand, board)
    assert quads.high_value == 4


def test_quads_low(quads):
    board = quads[1]
    hand = quads[0]
    quads = p.find_quads(hand, board)
    assert quads.low_value == 14


def test_not_quads(flush_one_pair):
    board = flush_one_pair[1]
    hand = flush_one_pair[0]
    quads = p.find_quads(hand, board)
    assert not quads


def test_not_boat(flush_one_pair):
    hand = flush_one_pair[0]
    board = flush_one_pair[1]
    boat = p.find_full_house(hand, board)
    assert not boat


def test_boat_high(boat):
    hand = boat[0]
    board = boat[1]
    boat = p.find_full_house(hand, board)
    assert boat.high_value == 14


def test_boat_low(boat):
    hand = boat[0]
    board = boat[1]
    boat = p.find_full_house(hand, board)
    assert boat.low_value == 4


def test_boat_type(boat):
    hand = boat[0]
    board = boat[1]
    boat = p.find_full_house(hand, board)
    assert boat.type == 'boat'



src/holdem_sim/poker_functions.py
import random
from collections import Counter
from dataclasses import dataclass

card_values = [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]
ranks = ['2', '3', '4', '5', '6', '7', '8', '9', 'T', 'J', 'Q', 'K', 'A']
rank_value = dict(zip(ranks, card_values))
value_rank = dict(zip(card_values, ranks))
suits = ['c', 'd', 'h', 's']
hand_values = {'hc': 1,
               'pair': 2,
               '2pair': 3,
               '3ok': 4,
               'straight': 5,
               'flush': 6,
               'boat': 7,
               '4ok': 8,
               'straight_flush': 9
               }

HAND_REGISTRY = []

#####     CLASSES     #####

@dataclass
class Card:
    def __init__(self, card_str):
        """
        Initializes a new instance of the Card class.

        Parameters:
            card_str (str): A string representing a card, with the rank as the first character and the suit as the
            second character.

        Returns:
            None
        """
        self.rank = str(card_str[0])
        self.suit = card_str[1]
        self.name = self.rank + self.suit
        self.value = rank_value[self.rank]

    def __str__(self):
        return self.name


    def __getitem__(self, item):
        if item == 'rank':
            return self.rank
        elif item == 'suit':
            return self.suit
        elif item == 'name':
            return self.name
        elif item == 'value':
            return self.value


@dataclass()
class Hand:
    def __init__(self, type, high_value, low_value = 0, kicker=0):
        """Type = name of hand (e.g. Pair)
        high_value = value.  either the high card in straight or flush, the set in full house, the top pair in 2pair, etc
        low_value = the lower pair in 2 pair or the pair in a full house
        kicker = value of the kicker in the hand.  Can be null
        """
        if kicker in card_values:
            kicker_rank = value_rank[kicker]
        else:
            kicker_rank = 0
        if low_value in card_values:
            low_rank = value_rank[low_value]
        else:
            low_rank = 0
        self.type = type
        self.hand_value = hand_values[type]
        self.kicker = kicker
        self.kicker_rank = kicker_rank
        self.high_value = high_value
        self.high_rank = value_rank[self.high_value]
        self.low_value = low_value
        self.low_rank = low_rank

    def __str__(self):
        return self.type + '-' + self.high_rank

    def __getitem__(self, item):
        if item == 'type':
            return self.type
        elif item == 'hand_value':
            return self.high_value
        elif item == 'kicker':
            return self.kicker
        elif item == 'kicker_rank':
            return self.kicker_rank
        elif item == 'high_value':
            return self.high_value
        elif item == 'high_rank':
            return self.high_rank
        elif item == 'low_value':
            return self.low_value
        elif item == 'low_rank':
            return self.low_rank


class Deck(list):
    def __init__(self, deck):
        self.deck = deck

    def __getitem__(self, item):
        return self.deck[item]

    def __iter__(self):
        for elem in self.deck:
            yield elem

    def __len__(self):
        return len(self.deck)

    def deal_card(self):
        """Select a random card from the deck.  Return the card and the deck with the card removed"""
        i = random.randint(0, len(self)-1)
        card = self[i]
        self.deck.pop(i)
        return card, self

    def update_deck(self, card):
        """Remove card from deck"""
        deck_names = [card.name for card in self.deck]
        if isinstance(card, Card):
            card_name = card.name
        else:
            card_name = card
        deck_idx = deck_names.index(card_name)
        self.deck.pop(deck_idx)


#####     USEFUL FUNCTIONS     #####

def register(func):
    """Add a function to the hand register"""
    HAND_REGISTRY.append(func)
    return func

def make_card(input_list: list):
    """Input_list is either a list of Card objects or string Objects.  If Cards, return the cards.
      If string, convert to Card and return"""
    if len(input_list) == 0:
        return input_list
    elif isinstance(input_list[0], Card):
        return input_list
    else:
        card_list = [Card(card) for card in input_list]
        return card_list


def generate_deck():
    """
    Generates a deck of cards.

    Returns:
        Deck: A deck of cards.
    """
    deck = []
    for rank in ranks:
        for suit in suits:
            card_str = rank + suit
            _card = Card(card_str)
            deck.append(_card)
    deck = Deck(deck)
    return deck


#####     POKER     #####
def find_multiple(hand, board, n=2):
    """
    Find a multiple of cards in the given hand and board.

    Args:
        hand (List[str]): The cards in the player's hand.
        board (List[str]): The cards on the board.
        n (int, optional): The number of cards to find a multiple of. Defaults to 2.

    Returns:
        Hand or bool: The multiple hand if found, False otherwise.
    """
    hand = make_card(hand)
    board = make_card(board)
    multiple = False
    multiple_hand = None
    total_hand = hand + board
    values = [card.value for card in total_hand]
    c = Counter(values)
    for value in set(values):
        if c[value] == 2 and n == 2:
            multiple = True
            hand_type = 'pair'
            high_value = value
            low_value = max([value for value in values if value != high_value])
            kicker = max([value for value in values if value not in [high_value, low_value]])
            multiple_hand = Hand(hand_type, high_value, low_value=low_value, kicker=kicker)
            return multiple_hand
        elif c[value] == 3 and n == 3:
            multiple = True
            hand_type = '3ok'
            high_value = value
            low_value = max([foo for foo in values if foo != high_value])
            kicker = max([bar for bar in values if bar not in [high_value, low_value]])
            multiple_hand = Hand(hand_type, high_value, low_value=low_value, kicker=kicker)
            return multiple_hand
        elif c[value] == 4 and n == 4:
            multiple = True
            hand_type = '4ok'
            high_value = value
            low_value = max([value for value in values if value != high_value])
            multiple_hand = Hand(hand_type, high_value, low_value=low_value)
            return multiple_hand
    return multiple


def evaluate_straight(values):
    """Evaluates a list of card values to determine whether there are 5 consecutive values"""
    straight = False
    count = 0
    straight_hand_values = []
    sranks = [bit for bit in reversed(range(2, 15))]
    sranks.append(14)
    for rank in sranks:
        if rank in values:
            count += 1
            straight_hand_values.append(rank)
            if count == 5:
                straight = True
                return straight, straight_hand_values
        else:
            count = 0
            straight_hand_values = []
    return straight, straight_hand_values


@register
def find_straight_flush(hand, board):
    """Find a straight flush in a given hand/board combination"""
    hand = make_card(hand)
    board = make_card(board)
    straight_flush = False
    flush = find_flush(hand, board)
    if flush:
        total_hand = hand + board
        total_hand = [card for card in total_hand]
        hand_suits = [card.suit for card in total_hand]
        c = Counter(hand_suits)
        flush_suit = c.most_common(1)[0][0]
        flush_hand = [card.value for card in total_hand if card.suit == flush_suit]
        straight_flush, straight_hand = evaluate_straight(flush_hand)
        if straight_flush:
            high_value = max(straight_hand)
            hand_type = 'straight_flush'
            straight_flush_hand = Hand(hand_type,high_value)
            return straight_flush_hand
        else:
            return straight_flush
    else:
        return straight_flush


@register
def find_quads(hand, board):
    quads = find_multiple(hand, board, n=4)
    return quads


@register
def find_full_house(hand, board):
    """Is there a full house?"""
    hand = make_card(hand)
    board = make_card(board)
    boat = False
    boat_hand = None
    total_hand = hand + board
    values = [card.value for card in total_hand]
    c = Counter(values)
    for value in set(values):
        if c[value] == 3:
            high_value = value
            c.pop(value)
            for value in set(values):
                if c[value] > 1:
                    low_value = value
                    kicker = max([value for value in values if value != high_value and value != low_value])
                    boat_hand = Hand('boat', high_value, low_value=low_value, kicker=kicker)
                    boat = True
                    return boat_hand
    return boat


@register
def find_flush(hand, board):
    """Does any combination of 5 cards in hand or on board amount to 5 of the same suit"""
    hand = make_card(hand)
    board = make_card(board)
    total_hand = hand + board
    total_hand_suits = [card.suit for card in total_hand]
    flush = False
    c = Counter(total_hand_suits)
    for suit in total_hand_suits:
        if c[suit] >= 5:
            flush = True
    if flush:
        flush_cards = [card for card in total_hand if card.suit == c.most_common(1)[0][0]]
        high_value = max([card.value for card in flush_cards])
        flush_hand = Hand('flush', high_value)
        return flush_hand
    else:
        return flush


@register
def find_straight(hand, board):
    """Find a straight in a given hand/board combination"""
    hand = make_card(hand)
    board = make_card(board)
    straight = False
    straight_hand = None
    high_value = 2
    reqd_hand_size = 5  # required hand size gives us some flexibility at the cost of more lines.  could be more efficient if we say 'if len(values)<5'
    total_hand = hand + board
    values = [*set(card.value for card in total_hand)]
    slices = len(values) - reqd_hand_size
    if slices < 0:
        return straight
    else:
        straight, straight_hand_values = evaluate_straight(values)
        if straight:
            hand_type = 'straight'
            if 14 in straight_hand_values:  # all([5,14]) does not work here so using nested ifs.
                if 5 in straight_hand_values:
                    high_value = 5
            else:
                high_value = max(straight_hand_values)
            straight_hand = Hand(hand_type, high_value)
            return straight_hand
        else:
            return straight


@register
def find_trips(hand, board):
    """
    Find Three-of-a-kind in a hand and board combination.

    Args:
        hand (list): The hand of cards.
        board (list): The cards on the board.

    Returns:
        list: A list of trips found in the hand and board combination.
    """
    trips = find_multiple(hand, board, n=3)
    return trips


@register
def find_two_pair(hand, board):
    """Is there two-pair?"""
    hand = make_card(hand)
    board = make_card(board)
    two_pair = False
    # two_pair_hand = None
    total_hand = hand + board
    values = [card.value for card in total_hand]
    c = Counter(values)
    for value in values:
        if c[value] > 1:
            pair1 = Hand('pair', value)
            c.pop(value)
            for value in values:
                if c[value] > 1:
                    pair2 = Hand('pair', value)
                    kicker = max([value for value in values if value != pair1.high_value and value != pair2.high_value])
                    two_pair_hand = Hand('2pair', max(pair1.high_value, pair2.high_value), low_value=min(pair1.high_value, pair2.high_value), kicker=kicker)
                    two_pair = True
                    return two_pair_hand
    return two_pair


@register
def find_pair(hand, board):
    pair = find_multiple(hand, board, n=2)
    return pair


@register
def find_high_card(hand, board):
    """
    A function to find the high card in a hand and board.

    Parameters:
    - hand (list): A list of cards representing the player's hand.
    - board (list): A list of cards representing the community board cards.

    Returns:
    - high_card_hand (Hand): An object representing the high card hand.

    This function takes a player's hand and the community board cards and finds the highest card value in the combined hand. It then creates a Hand object with the high card value as the main value, the second highest card value as the low value, and the third highest card value as the kicker. The function returns the created Hand object representing the high card hand.
    """
    hand = make_card(hand)
    board = make_card(board)
    total_hand = hand + board
    total_hand_values = [card.value for card in total_hand]
    total_hand_values.sort()
    high_value = total_hand_values[-1]
    low_value = total_hand_values[-2]
    kicker = total_hand_values[-3]
    high_card_hand = Hand('hc', high_value,low_value=low_value, kicker=kicker)
    return high_card_hand
